Add support in the rebalancer for the user to provide minimum number of serving replicas

siddharthteotia opened the pull request:
The current implementation of no-downtime rebalancing keeps up at least 1 serving replica up for each segment. Having only 1 serving replica up can potentially hurt throughput.

- The changes here add support for the user to specify the minimum number of serving replicas that should be kept up for no-downtime rebalancing.
- Rebalancer bails out if external view will never converge (ERROR state)
- Added unit tests in TableRebalancerTest for the new code.
- Added cluster integration test to cover all scenarios for more coverage and end-to-end testing
using entry point in tools package -- RebalanceTable command in pinot-admin.sh script

siddharthteotia commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -212,39 +214,87 @@ private IdealState getNextState(IdealState currentState, IdealState targetState,
+    int minReplicasToKeepUp = rebalanceUserConfig.getInt(RebalanceUserConfigConstants.MIN_REPLICAS_TO_KEEPUP_FOR_NODOWNTIME,
+        RebalanceUserConfigConstants.DEFAULT_MIN_REPLICAS_TO_KEEPUP_FOR_NODOWNTIME);
+    if (minReplicasToKeepUp >= currentIdealStateSegmentHosts.size()) {
+      // if the minimum number of serving replicas in rebalance config is
```
One thought here was that PinotTableRestletResource can first use an API of TableRebalancer to check if the user specified min replica number makes sense or not. TableRebalancer will look at ideal state and see the number of replicas for each segment. If the user specified value >= number of replicas then we throw an exception right there before rest api code submits the async rebalance task. 

Right now such an API is not implemented. Rather we take the user specified value as is and at the time of rebalancing if we found the value to be invalid, we default to 1

mcvsubbu replied:
I think it is better to error out ahead of time if the min replicas to keep alive is greater than or equal to the number of replicas of the table,

siddharthteotia replied:
The reason I didn't do it is that I thought number of replicas could be different for each segment. So to validate the user specified value, we have to go over the ideal state and for any segment if the condition is not met, we throw error. IdealState has a method getMinReplicaCount() and I am not sure if that can be used straight without iterating over all the segments.

Secondly, during the course of rebalancing, ideal state might change in ZK (due to some concurrent operation), and out version update will fail in that case and then we fetch the new ideal state and recompute target. So this requires that we still safeguard our code where we update one segment at a time to check for this condition as just checking once upfront upon entry into rebalancer code might not be enough.

I have now added a validation method we call upfront here before rebalancer does anything. I am wondering if there is any quick way to do so without going over the ideal state.

mcvsubbu replied:
You can use the Idealstate to get the number of replicas.

siddharthteotia replied:
done

siddharthteotia commented on pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/util/TableRebalancerTest.java:
```diff
@@ -27,14 +27,16 @@
 
 
-public class TableRebalancerTest {
+public class TableRebalancerTest extends ControllerTest {
```
This was by mistake -- need to clean up.

siddharthteotia replied:
done

siddharthteotia commented:
The latest changes have the following:

(1) Entry point in pinot-admin tool for table rebalancer
(2) Add a cluster integration test for the end-to-end path for the usage from pinot-admin tool

siddharthteotia commented:
cc @Jackie-Jiang , @snleee , @mcvsubbu

mcvsubbu commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -212,39 +218,132 @@ private IdealState getNextState(IdealState currentState, IdealState targetState,
+
+      if (!hostToRemove.equals("")) {
+        idealStateToUpdate.getInstanceStateMap(segmentId).remove(hostToRemove);
+        LOGGER.info("Removing host: {} for segment: {}", hostToRemove, segmentId);
```
this will be a per segment log. u want this in info?

siddharthteotia replied:
thought could be helpful for debugging -- this method has 3 debug level messages , one for entry in method and one each for adding and removing a host

mcvsubbu replied:
no, we can't have info logs per segment.

siddharthteotia replied:
Moved to debug

mcvsubbu commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -212,39 +218,132 @@ private IdealState getNextState(IdealState currentState, IdealState targetState,
+        }
+      }
+
+      if (!hostToAdd.equals("")) {
```
assert that it does not equal "", or, better, add a preconditions check

siddharthteotia replied:
done

mcvsubbu commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -212,39 +218,132 @@ private IdealState getNextState(IdealState currentState, IdealState targetState,
     } else {
       // remove one entry
-      idealState.getInstanceStateMap(segmentId).remove(srcMap.keySet().stream().findFirst().get());
+      String hostToRemove = "";
```
so the algorithm is to move one replica at a time until we reach a min number of replicas alive in target state, and then just move everything else. Would be good to note that in a comment above the method.

snleee replied:
+1

siddharthteotia replied:
updated javadoc

mcvsubbu commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -300,4 +399,55 @@ private void waitForStable(String resourceName)
+    }
+
+    /**
+     * Get the number of times we updated the ideal
```
this is really number of segments we updated, because we bump this once for each segment that we change in idealstate. We may then update the idealstate in one go for all the segments.  We should also keep the number of idealstate updates as a separate stat.

siddharthteotia replied:
changed the name of these stats to avoid confusion, also tracking the number of times we updated the ideal state in ZK (which happens after we have made updates to instance map of each segment)

mcvsubbu commented on pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/util/TableRebalancerTest.java:
```diff
@@ -117,4 +119,332 @@ public void downtimeUpdateWithNoCommonElements() {
+    // Essentially, host1 from current state got replaced with host4 from target state
+    // Now, what we are checking is that between toUpdate and target ideal state,
+    // there is exactly 1 host in common
+    verifyStateForMinReplicaConstraint(toUpdateIdealState, targetIdealState, 1, true);
```
verify stats?

siddharthteotia replied:
stats are verified further in the test

mcvsubbu commented on pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/TableRebalancerAdminToolClusterIntegrationTest.java:
```diff
@@ -0,0 +1,352 @@
+      throws Exception {
+    startZk();
+    startController();
+    startBroker();
```
You should not need broker for this test

siddharthteotia replied:
I am looking into it. But if I don't start broker, the table creation fails with HTTP bad request during the call to completeTableConfiguration that sends a POST request to controller for creating the new table.

mcvsubbu replied:
It could be that the table add specifies broker instances, and the controller checks if there are such instances.

snleee commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java:
```diff
@@ -472,7 +472,8 @@ public String rebalance(
       @ApiParam(value = "true|false") @DefaultValue("false") @QueryParam("includeConsuming") Boolean includeConsuming,
-      @ApiParam(value = "true|false") @DefaultValue("false") @QueryParam("downtime") Boolean downtime) {
+      @ApiParam(value = "true|false") @DefaultValue("false") @QueryParam("downtime") Boolean downtime,
+      @ApiParam(value = "1|2|3|4") @DefaultValue("1") @QueryParam("minReplicasToKeepUpForNoDowntime") Integer minReplicasToKeepUp) {
```
Instead of number, can you put the explanation on what this variable means? It looks like we need to choose one of 1-4.

siddharthteotia replied:
done

snleee commented on pinot-tools/src/main/java/org/apache/pinot/tools/admin/command/RebalanceCommand.java:
```diff
@@ -0,0 +1,114 @@
+  }
+
+  @Override
+  public void printUsage() {
```
Why are you trying to override `printUsage`?

I thought the framework would deal with printing information correctly (with the info that you wrote in `@Option` annotation

siddharthteotia replied:
printUsage() is overriden to also print some example usage of RebalanceCommand. So I let it print each option along with description by calling super.printUsage() and then print some sample usages.

mcvsubbu replied:
Can the --help option serve here?

siddharthteotia replied:
No that doesn't print example usage.

mcvsubbu replied:
Can you change the super method to call a printExamples() and printUsage()? That way other commands can also take advantage of this to print examples.  No better documentation than the one printed by the code!

siddharthteotia replied:
Done

snleee commented on pinot-tools/src/main/java/org/apache/pinot/tools/PinotTableRebalancer.java:
```diff
@@ -0,0 +1,123 @@
+        tableRebalancer.rebalance(tableConfig, rebalanceSegmentsStrategy, rebalanceUserConfig);
+
+    if (rebalanceResult.getStatus() == RebalanceResult.RebalanceStatus.FAILED) {
+      System.out.println("Failed to rebalance table: " + tableName);
```
Are we not using `LOGGER` in tools package?

siddharthteotia replied:
Initially didn't use it for manual testing. Switched to LOGGER now. It is being used in some of the other tools code as well.

siddharthteotia commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -162,10 +164,12 @@ public RebalanceResult rebalance(TableConfig tableConfig, RebalanceSegmentStrate
           waitForStable(tableNameWithType);
           // clear retries as it tracks failures with each idealstate update attempt
           retries = 0;
+          LOGGER.debug("External view converged for the change in ideal state. Will start the next iteration now");
```
done

siddharthteotia commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -162,10 +164,12 @@ public RebalanceResult rebalance(TableConfig tableConfig, RebalanceSegmentStrate
         if (zkBaseDataAccessor
             .set(idealStateKey.getPath(), nextIdealState.getRecord(), currentIdealState.getRecord().getVersion(),
                 AccessOption.PERSISTENT)) {
+          LOGGER.debug("Successfully persisted the ideal state in ZK. Will wait for External view to converge");
```
done

siddharthteotia commented on pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/TableRebalancerAdminToolClusterIntegrationTest.java:
```diff
@@ -0,0 +1,352 @@
+    public class FakeSegmentStateModel extends StateModel {
+
+      @Transition(from = "OFFLINE", to = "ONLINE")
+      public void onBecomeOnlineFromOffline(Message message, NotificationContext context) {
```
dont -- added a test to explicitly throw exception during state transition and check that rebalancer detects the error state in external view and bails out

siddharthteotia commented on pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/TableRebalancerAdminToolClusterIntegrationTest.java:
```diff
@@ -0,0 +1,352 @@
+        AccessOption.PERSISTENT);
+
+    // should be enough for the callbacks to come and go away
+    Thread.sleep(2000);
```
done -- using a new table in each test

siddharthteotia commented on pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/TableRebalancerAdminToolClusterIntegrationTest.java:
```diff
@@ -0,0 +1,352 @@
+    Assert.assertEquals(2, _stateTransitionStats.dropFromOff);
+
+    // clear for next test to avoid mixing state transition stats
+    clearIdealState(2);
```
done -- added a test

siddharthteotia commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -162,10 +164,12 @@ public RebalanceResult rebalance(TableConfig tableConfig, RebalanceSegmentStrate
+          LOGGER.debug("External view converged for the change in ideal state. Will start the next iteration now");
           continue;
         }
         // in case of any error, we retry a bounded number of types
```
done, added info before attempting CAS on ZK. also dumped the changed ideal state (the one that we will try to atomically write in ZK)

siddharthteotia commented:
Addressed review comments and added more tests.

siddharthteotia commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -257,47 +390,162 @@ private void setTargetState(IdealState idealState, String segmentId, Map<String,
+          break;
+        } else {
+          LOGGER.info("Waiting for externalView to match idealstate for table:" + resourceName);
+          Thread.sleep(EXTERNAL_VIEW_CHECK_INTERVAL_MS);
```
Although we now bail out on detecting error in external view, there is still a possibility of rebalancer waiting indefinitely for external view to converge simply because the changes haven't yet reflected and there is no error. I feel after updating ideal state in ZK when we come here to wait for stabilization, we should spend a fix amount of time in this method if isStable() never returns true

mcvsubbu replied:
agreed. Fixed, max time and the bail out.

siddharthteotia replied:
Done. using a max wait time of 5mins.

siddharthteotia commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -139,58 +145,88 @@ public RebalanceResult rebalance(TableConfig tableConfig, RebalanceSegmentStrate
+          // remove the cause as it is private and is only used to detect exact error within this class
+          throw new IllegalStateException(e.getMessage());
+        } else {
+          LOGGER.warn("Caught exception while or after updating ideal state for resource: {}", tableNameWithType, e);
```
It makes total sense to continue (restart by regetting the current ideal state from ZK and recomputing the target) if the atomic operation failed in ZK due to version mismatch. 

However, if there is any other unpredictable exception (other than Illegal state exception due to error state in external view), I suggest we still bail out

mcvsubbu replied:
agreed

siddharthteotia replied:
Done

siddharthteotia commented:
Ping.
All review comments have been addressed. Please take another look

mcvsubbu commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -139,58 +145,88 @@ public RebalanceResult rebalance(TableConfig tableConfig, RebalanceSegmentStrate
-        // in case of any error, we retry a bounded number of types
+        // in case of any error, we retry a bounded number of times
       } catch (ZkBadVersionException e) {
+        // we will go back in the loop and reattempt by recomputing the target ideal state
```
In the log in line 187, pls include the version you were expecting to see

siddharthteotia replied:
Done

mcvsubbu commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -212,39 +248,136 @@ private IdealState getNextState(IdealState currentState, IdealState targetState,
+      Map<String, String> targetIdealStateSegmentHosts, IdealState idealStateToUpdate,
+      Configuration rebalanceUserConfig) {
+
+    LOGGER.info("Will update instance map of segment: {}", segmentId);
```
i think this should move to debug since w will be logging once per segment. Tables can have 1000s of segments. All info logs in this method should turn into debug

siddharthteotia replied:
Done

mcvsubbu commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -257,47 +390,162 @@ private void setTargetState(IdealState idealState, String segmentId, Map<String,
+   * @return true if external view is same as ideal state, false otherwise
    */
-  public int isStable(String tableName) {
+  private boolean isStable(String tableName) {
```
If we return a count of number of segments converged, that can show a progress bar. I am fine if that is a subsequent PR

siddharthteotia replied:
Sounds good.

siddharthteotia replied:
Done in the same PR -- keeping track of the number of segments not converged and then printing a log message

mcvsubbu commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -257,47 +390,162 @@ private void setTargetState(IdealState idealState, String segmentId, Map<String,
+  public static class RebalancerStats {
+    private int dryRun;
+    private int updatestoIdealStateInZK;
+    private int directUpdatesToSegmentInstanceMap;
```
which one of these equals numSegmentsMoved?

siddharthteotia replied:
None of them -- I am deriving the num segments moved in the caller of the method that updates the instance map of each segment in a loop. Then I am also logging the difference between the updated ideal state and target

mcvsubbu commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -257,47 +390,162 @@ private void setTargetState(IdealState idealState, String segmentId, Map<String,
+    }
+
+    /**
+     * Get the number of times we updated the instance-state
```
I am not clear on what this means. Is it the max number of times we updated in one step?

siddharthteotia replied:
No, this is the total number of time we updated the host-state mapping of a segment to target mapping in a single step. This is cumulative across all segments.

siddharthteotia commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -139,58 +145,88 @@ public RebalanceResult rebalance(TableConfig tableConfig, RebalanceSegmentStrate
           waitForStable(tableNameWithType);
           // clear retries as it tracks failures with each idealstate update attempt
           retries = 0;
+          LOGGER.debug("External view converged for the change in ideal state for table {}. Will start the next iteration now",
```
Moved to info. We only check at the beginning of the while loop if next iteration is needed. I have changed the log message  to have "will start next iteration (if any)

siddharthteotia commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -139,58 +145,88 @@ public RebalanceResult rebalance(TableConfig tableConfig, RebalanceSegmentStrate
         if (zkBaseDataAccessor
             .set(idealStateKey.getPath(), nextIdealState.getRecord(), currentIdealState.getRecord().getVersion(),
                 AccessOption.PERSISTENT)) {
+          LOGGER.debug("Successfully persisted the ideal state for table {} in ZK. Will wait for External view to converge",
```
Done

siddharthteotia commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -110,6 +115,8 @@ public RebalanceResult rebalance(TableConfig tableConfig, RebalanceSegmentStrate
       result.setIdealStateMapping(idealState.getRecord().getMapFields());
       result.setPartitionAssignment(partitionAssignment);
+      result.setStatus(RebalanceResult.RebalanceStatus.DONE);
+      _rebalancerStats.dryRun = 1;
```
For a given invocation of Rebalancer tool (either through command line or rest api), we create a single instance of TableRebalancer and invoke the rebalance() method. The stats are computed in a cumulative manner for that invocation. However, if the user invoked it in dry run mode then dryRun is set to 1 -- it can only be 1 for that invocation

mcvsubbu replied:
my bad, sorry

siddharthteotia commented on pinot-common/src/main/java/org/apache/pinot/common/restlet/resources/RebalanceResult.java:
```diff
@@ -28,17 +28,19 @@
   private Map<String, Map<String, String>> idealStateMapping;
   private PartitionAssignment partitionAssignment;
-  private String status;
+  private RebalanceStatus status;
```
done

siddharthteotia commented:
Addressed latest review comments

mcvsubbu commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java:
```diff
@@ -472,7 +472,9 @@ public String rebalance(
-      @ApiParam(value = "true|false") @DefaultValue("false") @QueryParam("downtime") Boolean downtime) {
+      @ApiParam(value = "true|false") @DefaultValue("false") @QueryParam("downtime") Boolean downtime,
+      @ApiParam(value = "number of serving replicas to keep alive per segment while rebalancing in no downtime mode")
+      @DefaultValue("1") @QueryParam("minReplicasToKeepUpForNoDowntime") Integer minReplicasToKeepUp) {
```
Suggest "minAvailableReplicas". @snleee can you think of other possibly better param names?

snleee replied:
Sounds better than the current name :)

siddharthteotia replied:
Done.

snleee commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -139,112 +153,293 @@ public RebalanceResult rebalance(TableConfig tableConfig, RebalanceSegmentStrate
     }
   }
 
+  private void validateMinReplicas(final String tableNameWithType,
```
I think that checking replication factor from table config should be enough for validation.

siddharthteotia replied:
Done. No longer iterating over ideal state. Just using idealstate.getReplicas()

snleee replied:
I was mentioning `replication` and `numReplicasPerPartition` from table config. If you want to use `idealstate.getReplicas()`, I think that you should double check on how it gets updated. I think that we usually don't depend on that number.

siddharthteotia replied:
Initial thought was to use one of the methods from table config but @mcvsubbu suggested to use the ideal state method. I can see that setReplicas() method on ideal state is being invoked through PinotTableRestletResource -> PinotHelixResourceManager -> PinotTableIdealStateBuilder -> IdealStateBuilder. In my unit tests I see this happening every time a post request is sent to the controller for creating the table. In fact the value that is being passed down to IdealState.setReplicas() is coming from table config's SegmentValidationAndRetetionConfig.getReplication()

mcvsubbu replied:
Checking from table config is a bit involved, different for offline, hlc and llc. This check is done during table creation, and the idealstate is created with the right number of replicas. Since we are getting idealstate here anyway, it makes sense to check for replicas on the idealstate

snleee commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -139,112 +153,293 @@ public RebalanceResult rebalance(TableConfig tableConfig, RebalanceSegmentStrate
+          if (e.getCause() instanceof ExternalViewErrored) {
+            LOGGER.error("External view reported error for table {} after updating ideal state", tableNameWithType);
+          } else if (e.getCause() instanceof ExternalViewConvergeTimeout) {
+            LOGGER.error("Timedout while waiting for external view to converge for table {}", tableNameWithType);
```
`Timed out`

siddharthteotia replied:
done

snleee commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -139,112 +153,293 @@ public RebalanceResult rebalance(TableConfig tableConfig, RebalanceSegmentStrate
-        .getBoolean(RebalanceUserConfigConstants.DOWNTIME, RebalanceUserConfigConstants.DEFAULT_DOWNTIME)) {
-      setTargetState(idealState, segmentId, targetMap);
+    // dump additional detailed info for debugging
+    if (LOGGER.isDebugEnabled()) {
```
do we really need this check?

siddharthteotia replied:
I thought we can check upfront to avoid running unnecessary logic that is needed to pretty print map by using joiner.

snleee commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -257,47 +452,198 @@ private void setTargetState(IdealState idealState, String segmentId, Map<String,
+        continue;
+      }
+
+      if (LOGGER.isDebugEnabled()) {
```
Do we need this check?

siddharthteotia replied:
Same as before. prettyPrintMap() function first uses joiner to come up with a pretty print version of map. I thought we can avoid it all together upfront by checking if debug is enabled or not.

snleee commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -257,47 +452,198 @@ private void setTargetState(IdealState idealState, String segmentId, Map<String,
-          LOGGER.debug("Mismatch: segment" + segment + " server:" + server + " state:" + state);
-          numDiff = numDiff + 1;
+        if (!mapEV.containsKey(server)) {
+          LOGGER.info("Host-state mapping of segment {} doesn't yet have server {} in external view",
```
I think that we should log `debug` level only in this loop. You're logging `num segments * replication` lines of log. This can easily be equivalent to thousands of lines of logs.

siddharthteotia replied:
Only for the segments and servers where external view hasn't converge. Right?

snleee replied:
oh i missed the if check :)

snleee commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -257,47 +452,198 @@ private void setTargetState(IdealState idealState, String segmentId, Map<String,
+          final String stateInIdealState = mapIS.get(server);
+          final String stateInExternalView = mapEV.get(server);
+          if (!stateInIdealState.equalsIgnoreCase(stateInExternalView)) {
+            LOGGER.info("Host-state mapping of segment {} has state {} in external view and state {} in ideal state",
```
The same as above. I would prefer to use `debug` level.

siddharthteotia replied:
This log will be only for segment and server combination where state is not same as the one in ideal state. Is that fine in debug?

snleee commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -257,47 +452,198 @@ private void setTargetState(IdealState idealState, String segmentId, Map<String,
+          // external view not yet converged
+          stable = false;
+          converged = false;
+        } else if (mapEV.get(server).equalsIgnoreCase("error")) {
```
I think that we should add the validation to check if there's any `ERROR` state in the external view before running any rebalance logic. In that case, we know for sure that `error` state caught here is caused by the rebalance. Then, it makes sense to stop the rebalance.

siddharthteotia replied:
Done. I feel we should return with FAILEd status code from the rebalancer once the pre-check on the external view detects error state. The reason being that there is probably no gain in going forward with rebalancing if external view is already not fine to begin with.

siddharthteotia commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -139,112 +153,293 @@ public RebalanceResult rebalance(TableConfig tableConfig, RebalanceSegmentStrate
+        final Map<String, String> segmentHostsMap = currentMapFields.get(segment);
+        if (minReplicasToKeepUp >= segmentHostsMap.size()) {
+          final StringBuilder sb = new StringBuilder();
+          sb.append("User specified invalid number of min replicas: ").append(minReplicasToKeepUp)
```
Changed to use logger. My main reason for using string builder was to construct the message once and use it in both logger and exception

siddharthteotia commented:
Addressed review comments. Added integration tests for realtime LLC (consuming, without consuming), realtime HLC and segment movement stats.

@snleee  brought up a point w.r.t throwing exception on detecting ERROR state in external view.

Currently we throw exception on detecting error in external view. What are the pros and cons of ignoring it and continuing with the hope that probably there is 1 of many segments that is in ERROR state and may be we can continue with rebalancing v/s the current approach of aborting the rebalancer midway by throwing exception.

mcvsubbu commented on pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/TableRebalancerAdminToolClusterIntegrationTest.java:
```diff
@@ -0,0 +1,877 @@
+
+    try {
+      startFakeServers(NUM_INITIAL_SERVERS, CommonConstants.Helix.DEFAULT_SERVER_NETTY_PORT, true);
+      startKafka();
```
things may run faster if you startKafka and just keep it around and stopKafka() at the end. We don't consume any messages, so it is ok to just keep it running throughout.

siddharthteotia replied:
Done

siddharthteotia commented on pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/TableRebalancerAdminToolClusterIntegrationTest.java:
```diff
@@ -0,0 +1,877 @@
+      tableRebalancer.rebalance(tableName, "OFFLINE");
+      Assert.fail("Expecting exception");
+
+      final TableRebalancer.RebalancerStats stats = tableRebalancer.getRebalancerStats();
```
Good catch. Done.

siddharthteotia commented on pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/TableRebalancerAdminToolClusterIntegrationTest.java:
```diff
@@ -0,0 +1,877 @@
+      Assert.assertTrue(e.getMessage().contains("External view reports error state for segment segment0"));
+    }
+    _raiseErrorOnSegmentStateTransition = false;
+    stopFakeServers();
```
Done

siddharthteotia commented on pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/util/TableRebalancerTest.java:
```diff
@@ -117,4 +119,332 @@ public void downtimeUpdateWithNoCommonElements() {
   }
+
+  /**
+   * Test for now downtime rebalance with no common hosts between
```
Done

siddharthteotia commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java:
```diff
@@ -472,7 +472,9 @@ public String rebalance(
       @ApiParam(value = "true|false") @DefaultValue("false") @QueryParam("includeConsuming") Boolean includeConsuming,
-      @ApiParam(value = "true|false") @DefaultValue("false") @QueryParam("downtime") Boolean downtime) {
+      @ApiParam(value = "true|false") @DefaultValue("false") @QueryParam("downtime") Boolean downtime,
+      @ApiParam(value = "number of serving replicas to keep alive per segment while rebalancing in no downtime mode")
```
Done

siddharthteotia commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java:
```diff
@@ -472,7 +472,9 @@ public String rebalance(
       @ApiParam(value = "true|false") @Nonnull @DefaultValue("true") @QueryParam("dryrun") Boolean dryRun,
       @ApiParam(value = "true|false") @DefaultValue("false") @QueryParam("includeConsuming") Boolean includeConsuming,
-      @ApiParam(value = "true|false") @DefaultValue("false") @QueryParam("downtime") Boolean downtime) {
+      @ApiParam(value = "true|false") @DefaultValue("false") @QueryParam("downtime") Boolean downtime,
```
Done

siddharthteotia commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -67,15 +72,18 @@
 
   private static final int MAX_RETRIES = 10;
   private static final int EXTERNAL_VIEW_CHECK_INTERVAL_MS = 30000;
+  private static final int EXTERNVAL_VIEW_STABILIZATON_MAX_WAIT_MS = 300000;
```
Done. using limit of 20mins for now. We can refine this in follow-up PR

siddharthteotia commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -257,47 +451,208 @@ private void setTargetState(IdealState idealState, String segmentId, Map<String,
+
+    if (!done) {
+      LOGGER.error("External view didn't converge for table {} within max wait time of 5mins. Bailing out",
+          resourceName);
```
Done

siddharthteotia commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -257,47 +451,208 @@ private void setTargetState(IdealState idealState, String segmentId, Map<String,
+    }
+
+    /**
+     * Get the number of times we updated the instance-state
```
Not sure what that implies. I have updated the javadoc if it is more clear now. We can discuss this

siddharthteotia commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -257,47 +451,208 @@ private void setTargetState(IdealState idealState, String segmentId, Map<String,
+
+    /**
+     * Get the number of times we updated the instance-state
+     * mapping acrosss all segments in ideal state in one step. This
```
Not sure what that implies. I have updated the javadoc if it is more clear now. We can discuss this

siddharthteotia commented:
Addressed all review comments

mcvsubbu commented on pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/TableRebalancerAdminToolClusterIntegrationTest.java:
```diff
@@ -0,0 +1,868 @@
+        AccessOption.PERSISTENT);
+  }
+
+  public class FakeServerSegmentStateModelFactory extends StateModelFactory<StateModel> {
```
would be a good thing to re-use methods to add/stop fake instances in ControllerTest (jackie cleaned up the tests recently), but we can leave that to a later PR. ControllerTest is a super-class of TableRebalancerAdminToolClusterIntegrationTest
Can you please add a TODO here?

siddharthteotia replied:
Done

siddharthteotia commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -124,13 +123,32 @@ public RebalanceResult rebalance(TableConfig tableConfig, RebalanceSegmentStrate
       return result;
     }
 
+    // before running rebalancer if external view is fine to begin with
```
Moved it to before dry run

siddharthteotia commented on pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/util/TableRebalancerTest.java:
```diff
@@ -112,9 +106,322 @@ public void downtimeUpdateWithNoCommonElements() {
+
+  private void verifyStateForMinReplicaConstraint(
+      final IdealState updated, final IdealState target,
+      final int same, final boolean sizeCheck) {
```
Done

siddharthteotia commented on pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/util/TableRebalancerTest.java:
```diff
@@ -112,9 +106,322 @@ public void downtimeUpdateWithNoCommonElements() {
+   * number of minimum replicas up while rebalancing
+   */
+  @Test
+  public void noDowntimeUpdateWithNoCommonHostsAndInvalidMinReplicas() {
```
This test file doesn't exercise the full code in TableRebalancer. It was written that way originally. IT directly calls the updateSegment() method that just updates the ideal state -- the check is still done there to safeguard code.

mcvsubbu replied:
ok, i think it is good to have that code with the minreplicas for possible race conditions when 2 of these are run, or a background and foreground versions are run together.

siddharthteotia commented:
Addressed latest review comments

mcvsubbu commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java:
```diff
@@ -139,112 +171,293 @@ public RebalanceResult rebalance(TableConfig tableConfig, RebalanceSegmentStrate
-      LOGGER.debug("Segment " + segmentId + " has common entries between current and expected ideal state");
-      setTargetState(idealState, segmentId, targetMap);
+    // we are in no-downtime mode
+    if (minAvailableReplicas >= currentIdealStateSegmentHosts.size()) {
```
Not sure what you mean by replicas going down .. we bail out if we see ERROR state in externalview.  We look at the idealstate for numreplicas, so if something changes in the tableconfig we are agnostic. we don't (yet) support the scenario of multiple instances of rebalancer tool running

siddharthteotia replied:
Discussed offline that is fine to keep the check

siddharthteotia commented:
Thanks @snleee , @mcvsubbu for reviewing and approving the changes.

Some follow up items I will address in upcoming PRs:

1 - Revisit the logic where we fetch the ideal state again and recompute the target. There is potential room for optimization here  -- fetch the ideal state again and recompute target only if we previously failed to write in ZK due to version mismatch.

2. For the rest api based access that runs rebalance in async mode, we need to show which controller is responding to the rebalance request.

3. Re-use the recently merged controller test apis in the integration test to start/stop fake servers.

4. See if logging can be improved/changed to ensure info level messages are giving enough information.
